Simulator Log View

eurocoinsalbum opened the pull request:
basic table which shows the command which have been sent to the simulator.
Can be extended later on to have a link to the block which initiated the command.
Didn't manage to remove the whitespace in that table
Added react-table module

ginonott commented:
I think including a table module might be a little over kill. You can render a table like this in React:
```
const data = [{foo: ''} ... ];

return (
  <table>
    <thead>
      <tr><th>Foo Column Header</th></tr>
    </thead>
    <tbody>
      {data.map(d => <tr><td>{d.foo}</td></tr>)} // maps data array to list of react "tr" components
    </tbody>
  </table>
)
```

ginonott commented on src/Views/SimulatorLog/SimulatorLog.tsx:
```diff
@@ -0,0 +1,57 @@
+export const SimulatorLog: FunctionComponent = () => {
+  const [tableData, setTableData] = useState<row[]>([]);
+  const motorPower0 = useSelector(getMotorPower(0));
+  const motorPower1 = useSelector(getMotorPower(1));
```
you are right that this isn't really scalable. Instead of building the table in this component, we can add an event log to the robotSimulatorSlice. Then retrieve the list of events with a selector.

It could look something like this:
```
interface IRobotSimulatorState {
    motors: {
      [channel: number]: number;
    };
    eventLog: Array<{eventType: "setPower", /* whatever other values you want ,*/ eventSource: "blockId" /* undefined if set via UI? */}>;
  }
  
  /**
   * Reducers for handling the state of the simulator such as whether it is executing or not.
   */
  export const robotSimulatorSlice = createSlice({
    initialState: {
      motors: {},
      eventLog: []
    } as IRobotSimulatorState,
    name: "simulator",
    reducers: {
      setPower(state, action: PayloadAction<{ channel: number; power: number, source?: string }>) {
        state.motors[action.payload.channel] = action.payload.power;
        state.eventLog.push({eventType: "setPower", eventSource: action.payload.source})
  
        return state;
      },
    },
  });
```

JamesWP commented:
Does the ui framework we have in the project have a table component built in?

I think we’re using Microsoft’s fluent ui.

Maybe take a look at this?
https://developer.microsoft.com/en-us/fluentui#/controls/web/detailslist

eurocoinsalbum commented:
I don't think that the RobotSimulatorSlice should keep a history of the commands. It's not his job and why should he? If he cares about other view's use-cases he will become too big.
In the log view I also want to show multiple logs. Like from the past n runs. Don't want to put all this burden on the RobotSimulatorSlice.
The other points mentioned I will refactor my code. Thanks for reviewing so fast!

JamesWP commented:
Redux is designed to store all state, idea idea being without it state is stored in many different places split up in different areas of the application. It becomes very hard to manage.

So if you think separating the log state into different parts is a good idea I’d suggest creating another slice for the state instead of storing it in the component itself.

ginonott commented:
That makes sense. Would it make sense to simply create a new redux slice
for a log command? We generally don't want react components to keep state
like event history.

On Sun, Jun 21, 2020, 11:22 AM JamesWP <notifications@github.com> wrote:

> Redux is designed to store all state, idea idea being without it state is
> stored in many different places split up in different areas of the
> application. It becomes very hard to manage.
>
> So if you think separating the log state into different parts is a good
> idea I’d suggest creating another slice for the state instead of storing it
> in the component itself.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/FRUK-Simulator/Simulator/pull/61#issuecomment-647141965>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEZSQPRCYP4AEVSZLS6MF2DRXYQRXANCNFSM4OD3MZKA>
> .
>

ginonott commented:
> So if you think separating the log state into different parts is a good idea I’d suggest creating another slice for the state instead of storing it in the component itself.

React components can also become unmounted and destroyed, losing all state and this seems like a persistent set of data so its definitely better to hoist it up into redux - but I definitely agree the robotSimulator might not be the best place to put it. Maybe a "monitorSlice" or something?

Here is an example of a reducer that "listens" to actions that belong to a different slice of state: https://github.com/FRUK-Simulator/Simulator/blob/master/src/BlocklyInterface/blocklySlice.ts#L29

In this case, you could add cases to your reducer for each of the robotSimulator commands

eurocoinsalbum commented:
ok, I think I got the point regarding moving the state into it's own slice away from the component. I will move it to ExecutionSlice which goes into it's own directory.
I just wonder how this would scale if I have 10000 commands and add another one. Will the entire content of the table be reacreated? The listener of the state change can't know which item has been added/changed and therefor needs to recreate all rows from scratch.

JamesWP commented:
React only create the elements on screen if they need to be created, if the state has one item appended to it then yes the state will be copied momentarily as the master copy is updated.  But react only creates one new HTML element for the new row.

But maybe there would be a limit where we would need to remove old things but I think it’s a problem for future us to solve :D

ginonott commented:
>  Will the entire content of the table be reacreated?

React has a way of dealing with this as well to determine what in a list needs to be updated in the actual DOM. See https://reactjs.org/docs/lists-and-keys.html#keys. As long as each element in the list

eurocoinsalbum commented:
thanks for the explanations.
I update the PR and pulled the state out of the component and put it into it's own slice.
Also removed react-table and replaced with simple HTML table.

ginonott commented on src/Views/SimulatorLog/SimulatorLog.tsx:
```diff
@@ -0,0 +1,48 @@
+  return (
+    <table>
+      <thead
+        style={{
```
Can you move the inline styles to `SimulatorLog.css`? I don't think you need to do "display" here.

eurocoinsalbum replied:
moved to css but whenever I remove display the UI looks weird

ginonott commented on src/Views/SimulatorLog/SimulatorLog.tsx:
```diff
@@ -0,0 +1,48 @@
+      </thead>
+      <tbody
+        style={{
+          display: "block",
```
can you move this to the `SimulatorLog.css` file? Also, I don't think you should be overriding the display of the tbody unless you had a specific reason for it?

eurocoinsalbum replied:
moved to css but whenever I remove display the UI looks weird

eurocoinsalbum commented:
Implemented all required changes from last comment. Thanks a lot - I learned again much :-)

eurocoinsalbum commented on src/Views/SimulatorLog/simulatorLogSlice.ts:
```diff
@@ -0,0 +1,53 @@
+      }),
+});
+
+export const getCurrentExecution = () => (state: RootState) => state.logs;
```
renamed to getCurrentExecutionLogs in order to indicate that those are for the current/last execution only. Long term we can have more execution logs like from the past n executions.

eurocoinsalbum commented:
edited some comments just right now and uploaded another change

JamesWP commented:
Two things, the log view should show the most recently executed action, otherwise its just showing stale data always, either reverse the order or scroll to show the last item when you add another one.
i also just noticed that it mentions 'percent' but the value goes between -1 and 1 not -100 and 100.

eurocoinsalbum commented:
1. new commands will now be added at the top.
2. I normalized the range to use -100 to 100 in the UI and -1 to 1 internally
